applications/inkjet.md

```diff
@@ -0,0 +1,280 @@
+
+This publication fine-tunes a GPT-J model on 2 million smart contracts, and tests if performance increases on writing code without security issues. While their methodology involved directly fine-tuning the model and updating the weights, and ours focuses on adding context through retrieval, the underlying concept of providing learning examples to improve performance remains similar. After the original fine-tuning, they find that insecure code was an issue in up to 70% of generations. After additional fine-tuning on vulnerable examples with vulnerability-constrained decoding, they were able to avoid insecure code generation up to 67% of the time. 
+
+We have taken these findings into consideration, and will ensure that the contracts in our dataset are functional. CoinFabrik Scout will be utilized to mitigate vulnerabilities. Use of this tool prevents the need for cross-contamination of our functional smart contracts with vulnerable ones in the vectorstore, the separation of which would have required heavy manual code annotations and metadata tagging.
+
+In terms of related work, we have [ongoing work](https://github.com/yu-jeffy/audit.me) studying the efficacy of vulnerability testing through RAG integrated LLMs on Ethereum Solidity smart contracts. In this study, we built a LangChain RAG-LLM pipeline, and created a vectorstore of 830 vulnerable Solidity smart contracts for retrieval. Results were promising, showing a 1.5x increased efficacy compared to [current literature](https://arxiv.org/abs/2306.12338). Using this as a proof of concept of RAG-LLM with smart contract data, we look to rebuild the pipeline towards the use case of authoring ink! smart contracts. We will construct a new pipeline from scratch, utilize LlamaIndex alongside LangChain, change the data processing and embedding methodology, and implement a much broader ink! smart contract dataset.
+
+## Development Roadmap :nut_and_bolt:
+
+### Overview
+
+- **Total Estimated Duration:** 5 months
+- **Full-Time Equivalent (FTE):** 3 FTE
+- **Total Costs:** 30,000 USD
+
+### Milestone 1 — Core RAG-LLM Functionality
+
+- **Estimated Duration:** 1 month
+- **FTE:**  3
+- **Costs:** 6,000 USD
+
+|  Number  | Deliverable | Specification |
+| -------: | ----------- | ------------- |
+|  **0a.** | License | Open-sourced under Apache 2.0. |
+|  **0b.** | Documentation | Code comments. Documentation for the prototype architecture and the setup process, explaining how a user can run a local instance of the prototype RAG system with our initial data. |
+|  **0c.** | Testing Guide | Unit tests will be provided for the prototype, with a testing guide explaining how to run them. |
+|  **0d.** | Docker | N/A |
+|       1. | Initial Prototype | Development of a basic LlamaIndex RAG system prototype integrated with `GPT-4`, using sentence embeddings. User can interact with the pipeline through the command line, interfacing with `GPT-4` with fetched documents from `Milvus and Weaviate`|
```

Comment:
I see a lot of GPT-related dependencies here. I just want to emphasise that any deliverable should be open-source and must NOT depend on any non-open-source technology. Quote from our [README](https://github.com/w3f/Grants-Program/blob/master/README.md):

> All code produced as part of a grant must be open-sourced, and it must also not rely on closed-source software for full functionality. We prefer Apache 2.0, but GPLv3, MIT, or Unlicense are also acceptable.

Are there any proprietary components involved with this project?

Reply:
We are interfacing with the OpenAI API to utilize GPT-4. The rest of our pipeline, such as the dataset and vectorstore are open sourced. Any code interfacing with the OpenAI API will be open source. 

Would the usage of GPT-4 be an issue?